File: src/main/java/com/hileco/cortex/instructions/ProgramException.java
package com.hileco.cortex.instructions;

import com.hileco.cortex.vm.ProgramContext;
import lombok.EqualsAndHashCode;
import lombok.Value;

@Value
@EqualsAndHashCode(callSuper = true)
public class ProgramException extends Exception {

    private final ProgramContext programContext;
    private final Reason reason;

    @Override
    public String getMessage() {
        return String.format("%s: %s", this.reason.toString(), this.programContext.toString());
    }

    public enum Reason {
        JUMP_TO_ILLEGAL_INSTRUCTION,
        JUMP_OUT_OF_BOUNDS,
        INSTRUCTION_LIMIT_REACHED_ON_PROCESS_LEVEL,
        INSTRUCTION_LIMIT_REACHED_ON_PROGRAM_LEVEL,
        STACK_LIMIT_REACHED,
        RETURN_DATA_TOO_LARGE,
        STACK_TOO_FEW_ELEMENTS,
        CALL_RECIPIENT_MISSING,
        WINNER
    }
}

File: src/main/java/com/hileco/cortex/instructions/ProgramException.kt
package com.hileco.cortex.instructions

class ProgramException(val reason: Reason) : Exception() {
    override val message: String?
        get() = "$reason"

    enum class Reason {
        JUMP_TO_ILLEGAL_INSTRUCTION,
        JUMP_TO_OUT_OF_BOUNDS,
        STORAGE_ACCESS_OUT_OF_BOUNDS,
        REACHED_LIMIT_INSTRUCTIONS_ON_VIRTUAL_MACHINE,
        REACHED_LIMIT_INSTRUCTIONS_ON_PROGRAM,
        STACK_OVERFLOW,
        STACK_UNDERFLOW,
        CALL_RETURN_DATA_TOO_LARGE,
        CALL_RECIPIENT_MISSING,
        WINNER,
        UNKNOWN_INSTRUCTION
    }
}
